runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricsContainerImpl.java

```diff
@@ -411,20 +412,23 @@ public int hashCode() {
   }
 
   /**
-   * Match a MetricName with a given namespace and a name. If the namespace or the name is null, it
-   * will be ignored for the match.
+   * Match a MetricName with a given metric filter. If the metric filter is null, the method always
+   * returns true.
    */
   private boolean matchMetricName(
-      MetricName metricName, @Nullable String namespace, @Nullable String name) {
-    return (namespace == null || namespace.equals(metricName.getNamespace()))
-        && (name == null || name.equals(metricName.getName()));
+      MetricName metricName, @Nullable Set<KV<String, String>> metricFilter) {
```

Comment:
HashSet<MetricName> metricFilter
I think it would make more sense to make the metricFilter just be a Set of MetricNames as well.  MetricName has a hash code so you can do efficient set lookups this way as well. It is used with Map types as a key normally, so this should work.

Also, make the metric  you are defining uses a MonitoringInfoMetricName, which doesn't have a name+namespace concept, it instead has a urn and labels. Instantiate a MonitoringInfoMetricName for it using the full URN and labels. Don't parse the URN and put it in a MetricName as a namespace+name

https://github.com/apache/beam/blob/a72460272354747a54449358f5df414be4b6d72c/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MonitoringInfoMetricName.java#L105

Reply:
Done. Using MetricName instead.